### ISSUE TITLE
Can't debug if script path contains non-ASCII character

### ISSUE THREAD

**Issue by robertlzj** [pkulchenko/ZeroBraneStudio#559]
> Program 'lua.exe' started in '...\ZeroBraneStudio\cfg' (pid: 9656).
> Can't start debugging for 'G:\□\test.lua'. Compilation error:
> Cannot open file G:\□\test.lua
> Program stopped (pid: 9656).

Script could be executed anyway.

Hope it will be easy to fix. : )

**Comment by pkulchenko** [comment]
@robertlzj, unfortunately this is caused by Windows using UTF-16 for some of the names and Lua using ASCII versions of functions to manipulate files. For details see [1](http://lua.2524044.n2.nabble.com/Support-for-Windows-unicode-paths-td3303722.html) and [2](http://lua-users.org/lists/lua-l/2012-08/msg00150.html).

The IDE uses a workaround to generate short names that can be used by Lua functions, but the short name generation can be disabled on some of the volumes, which is probably what happened in your case. You can check the current setting by using the following command: `fsutil 8dot3name query g:`. You may see something like: 

```
The volume state is: 1 (8dot3 name creation is disabled).
The registry state is: 2 (Per volume setting - the default).

Based on the above two settings, 8dot3 name creation is disabled on g:
```

If that's the case, you'd need to enable short name generation by using `fsutil 8dot3name set g: 0`. Unfortunately this is not enough, as you'll need to re-create the folder to regenerate its shortname (or use `fsutil file setshortname` command). See [this link for details and examples of these commands](http://blogs.msdn.com/b/winsdk/archive/2013/10/09/getshortpathname-doesn-t-return-short-path-name.aspx).

After this is done, the running/debugging from the IDE should work as expected. I'll probably add a warning message to this effect to the interpreter.

**Comment by robertlzj** [comment]
Thank you for the tip. I have a check, the `8dot3 name creation is enabled on g:` (which is my system volume).

I watch the path in ZeroBrane stuido, found inconsistency. I use `□` to represent non-ASCII character, and use `..` to simplify path, they are all showed correctly in output window.

1. Problem only happens when debug.
  
  > Program starting as '"...\ZeroBraneStudio\bin\lua.exe" -e "io.stdout:setvbuf('no')" "G:\Users\ADMINI~1\AppData\Local\Temp\.2DF7.tmp"'. 

  Using *8dot3* name and temp folder temp file.
  > Program 'lua.exe' started in 'G:\...\ZeroBraneStudio\cfg' (pid: 12784).
> Can't start debugging for 'G:\□\..\directory.lua'. Compilation error:
> Cannot open file G:\□\..\directory.lua

  Using *original* name.
  
2. When execute normally.
> Program starting as '"G:\..\ZeroBraneStudio\bin\lua.exe" -e "io.stdout:setvbuf('no')" "G:\□\..\DIRECT~1.LUA"'.

  Using *8dot3* name.

So, ..., any suggetion? :sweat:

**Comment by pkulchenko** [comment]
@robertlzj, you'll need to recreate that folder after you made the change to the `8dot3` setting. You can check if there is a short name assigned to the file/folder by running `dir /X` command. If there is no short name (it would be shown in the column right before the regular name), then it's not going to work.

**Comment by robertlzj** [comment]
I learn a little about `dir /x` and `8dot3` from your link also from others.
I check the path again on each related folder level.
It seems only names longer than 8 ASCII characters (**long name**) would need to convert to **short name**. Which, simple non-ASCII names won't need short name. So, `dir /x` will show this
>                  dialog4.lua
    DIRECT~1.LUA directory.lua


I simplify the path to test. `G:\Ｆ\test.lua` can be executed from zeroBrane studio (how can I write it for short with respect : ) ), but can't debug. Which sciTE4Lua works in both.

So :worried:

**Comment by pkulchenko** [comment]
> Because I found Lua5.2 interpreter can not even run script with this path.

Right; Lua is using Windows functions that can't deal with Unicode names (see the discussions I referenced earlier).

>              Ｘ           <- non-ASCII folder name

So, if it shows no short name for that folder, you won't be able to debug that script as it needs to be referenced in `dofile` command. To create a short name for it you either need to recreate it (as long as the short names are enabled) or create a name directly using `fsutil file setshortname` command.

**Comment by robertlzj** [comment]
Eh...I have test a lot a lot and found these:
- **short name**(accord with *8dot3*) is used *only(?)* when original path (folder / file) name is longger than 8 byte character.
- **create** or **rename** a folder / file will make a short name automaticly *only(?)* if the original name is short enough.
- There is no relevance between **non-ASCII** and **short name**.  
Means, non-ASCII contained folder / file names may not need a short name, and system won't generate it automaticly when *create* nor *rename*, if it is short enough.
- To execute file, the non-ASCII paths is harmless.  
Since it pass the file object, not the file path I think.(See test below)
(Thus, I think it's not necessary to pass file to interpreter by file's short name.)
- `dofile` can't handle non-ASCII contained path.  
And can handle long ASCII-consist path (even delete its short name).
- The path comes from original name or short name.  
**Only** when the original name is *long* enough, then the short name is used.  
Means, even set a short name manually, it may not be used. Like `x.lua` is a short name for `Ｘ.lua`, but won't be used.
- And I checked the 8dot3 name creation is enabled all time.

So, if original name is `Ｘ.lua`, it can't be debuged any way. And, if original name is long enough and its short name don't contain non-ASCII characters, then it could be debuged.

If original name is long enough and the short name is exist, the editor will choose short name, otherwise the original one.
I don't know if it is the editor's strategy or the system.

Some **test** a with a lot of mess write before, could be skipped as the result already said above.
> I delete the lua file‘s short name (using `fsutil` and checked using `dir /x`) of an original long name which contain ASCII-character or not, the lua file could be executed correctly.  
> Could see editor use short name if existed, or use originall name if not.
> > ..lua.exe" -e "io.stdout:setvbuf('no')" "G:\ＸＸＸ~1.LUA"'
> > ..lua.exe" -e "io.stdout:setvbuf('no')" "G:\ＸＸＸＸＸＸ.LUA"'
>
>  Even with short name, if there is non-ASCII character in it (like `"G:\ＸＸＸ~1.LUA"`), it can't debug. (But `"G:\123456~1.LUA"` can)
> I think 8dot3 name is not necessary since file could still be executed, debug, and `dofile` after  I delete short name.
> The short name for 'Ｘ.lua' is still 'Ｘ.lua' (in fact, in this case, short name is not necessary and won't be made automaticly). Which can't be handled by Lua `dofile`.

Last, sciTE4Lua could handle path like `G:\ｘ\ｘ.lua`. : )

**Comment by jjvbsag** [comment]
Yes, windows sucks. (At least in this case)
### Announcement
Just to let you know: I had similar problems (need to open guitar-pro files with Cyrillic and Japanese characters in the filename from lua) and wrote some extensions. These are to be used as
```lua
require"lfsutf8" lfs=lfsutf8
require"osutf8"  os=osutf8
require"ioutf8"  io=ioutf8
```
They are still test-in-progress, but seems to work fine for me. After above code all file-related functions like io.open, lfs.dir,... are mapped to use UTF8 strings. NOT fixed is dofile and passing lua scripts with UTF8 path.

When I believe they are functional, I may integrate them to my lua interpreter and to activate them with an additional parameter -u. I will gladly share a patch. (all done in lua5.1)

**Comment by pkulchenko** [comment]
@robertlzj, you are correct that running and debugging is done a bit differently. I pushed couple of changes that should (1) provide a warning and (2) improve debugging for Unicode names.

@jjvbsag is correct that it may need to be patched in Lua (and there are several patches mentioned in  the discussions I referenced earlier), but it's not an option in my case as I'd need to do this for all supported interpreters (including LuaJIT) and have it working for `io.*` and `dofile` (which is out of scope for my work).

In general you have several options for this:
1. patch Lua or load modules that handle Unicode names
2. use short names when possible; this should work with the IDE
3. avoid Unicode names

With the changes I pushed, the debugging should work even for Unicode names, but the "running" may not (as it needs to pass the filename as the parameter) and the things may still break somewhere in your script because of inability to open the file by Lua interpreter.
